Refactor text activity position handling

tobous opened the pull request:
Continuation of #862.

Migrates text edit activities and Jupiter operations to use the new line and in-line offset based positioning scheme introduced in #862.

With this PR, the migration to the new positioning scheme should be complete.

This migration to the new positioning scheme is done to make the position handling offset independent. This is a necessary step on the way towards internally only working on normalized content, which in turn is a step on the way towards providing IDE cross-compatibility.

Even though the main work is done in one commit (which I could not sensibly partition), it is still advisable to review the commits of this PR separately.

The leading commits of this PR add the necessary setup for the migration. The actual refactoring takes place in `[INTERNAL] Replace old positioning scheme for text edit activities`. I had a hard time further splitting up the refactoring commit. I wanted to create the new TextEditActivity and operation implementations separately, but as they are tightly integrated into the surrounding logic, this proved difficult. As this handling (i.e. renaming the existing logic to Old*, and adding the new implementations) would have extended the relative diff by multiple hundred lines and is very cumbersome to create retroactively, I decided against it.

~~This PR is still a work in progress as I still need to extend the following classes to include tests that cover the new, more complex cases (i.e. handling multi-line changes).~~
- ~~`SplitOperationTest`~~
- ~~`InclusionTest`~~
- ~~`GOTOInclusionTest`~~

~~Otherwise, this PR is ready to review.~~

Test suite overhauls for split operations and the inclusion transformation are done. The PR is ready to be reviewed.

### Commits

<details><summary><b>[INTERNAL] Adjust logic in EditorAPI to handle single positions</b></summary>
<br>

Adjusts the text position handling in EditorAPI to offer methods to
convert single positions/offsets.

Refactors the existing logic to use the newly added single
position/offset logic to avoid code duplication.

To keep the performance optimization that can be gained when converting
values that are located in the same line, there is an internal version
of  the single offset conversion methods that also accepts and returns
the needed values for such optimizations.

</details>

<details><summary><b>[INTERNAL][CORE] Make TextPosition comparable</b></summary>
<br>

Lets TextPosition extend 'Comparable<TextPosition>' to allow for an easy
way to compare text positions. This functionality will be useful for the
upcoming text edit activity refactoring to migrate it to also use text
positions.

</details>

<details><summary><b>[INTERNAL][CORE] Add TextPositionUtility</b></summary>
<br>

Adds TextPositionUtility, a utility class offering methods to calculate
text position related data.

These utility methods will be needed for the upcoming refactoring
migrating text edit activities and operations to use text positions
instead of offsets.

Also adds tests for TestPositionUtility.

</details>

<details><summary><b>[FIX][CORE] Add missed GotoInclusionTransformation.transformIndex case</b></summary>
<br>

Adds the case that the position is located in the area of the deletion,
in which case the new index should be the start of the deletion.

</details>

<details><summary><b>[JUNIT][CORE] Add OperationHelper test utility</b></summary>
<br>

Adds the OperationHelper class offering ease of use methods to work with
text activities and text operations in tests.

</details>

<details><summary><b>[JUNIT][CORE] Simplify Jupiter test cases using OperationHelper</b></summary>
<br>

Simplifies the Jupiter test logic by using the OperationHelper.

Removes unnecessary declarations of 'throws Exception' in the modified
tests.

</details>

<details><summary><b>[INTERNAL] Replace old positioning scheme for text edit activities</b></summary>
<br>

Replaces the old offset based positioning scheme with the new position
based scheme for text edit activities and Jupiter operations.

This migration was done as part towards the move to normalizing text
content internally.

Adjusts TextEditActivity, DeleteOperation, and InsertOperation to work
with text positions and line and offset deltas of the text changes.

Adjusts the integration of the changed components to use the new
positioning handling.

Cleans up the GotoInclusionTransformation and SplitOperation combination
logic by extracting logic that was used multiple times. Also comments
the code where deemed necessary.

Adjusts the test cases for the migrated logic to work with the new
positioning scheme. The existing test cases for the
GOTOInclusionTransformation and SplitOperation combination will have to
be extended to cover the newly added cases.

Extends the Document test helper class by an operation delta check. This
check ensures that all operations applied as part of a test actually
have the correct line and offset delta. This additional check is
important as incorrect delta values won't lead to a wrong document state
when applied but will lead to issues when further transforming the
operation or using it to transform other operations. As such further
transformations are not always done as part of the testing setup, it is
important to explicitly check for such issues.

</details>

<details><summary><b>[JUNIT][CORE] Add missing base test case for SplitOperationTest</b></summary>
<br>

Adds a missing test case for the base split operation combination using
only one-line changes.

With this test, all combination paths using only one-line changes are
tested.

</details>

<details><summary><b>[NOP][E] Remove unused EditorAPI method</b></summary>
<br>

Removes a now unused EditorAPI method added by the selection handling
refactoring.

The method is no longer necessary as the text edit activities now also
use the new positioning scheme.

</details>

<details><summary><b>[JUNIT][CORE] Add additional test cases for split operation combining</b></summary>
<br>

Adds additional multi-line test cases for the logic combining split
operations.

Adds additional needed test utility methods to OperationHelper.

Renames LINE_SEPARATOR to EOL in OperationHelper to keep the test lines
shorter.

</details>

<details><summary><b>[JUNIT][CORE] Add missing case not covered by base transformation test</b></summary>
<br>

Adds a missing base test case that was not covered by the previous
GOTOInclusionTransformationTest.

Adds the case where the deletion operation to transform with is complete
contained in the deletion information to transform.

With this, all base cases not containing line breaks should be covered.

</details>

<details><summary><b>[JUNIT][CORE] Add additional test cases for multi line transformations</b></summary>
<br>

Extends the InclusionTransformationTest test suite. Adds test for all
possible operation transformations using single- and multi-line content
changes.

The added tests explicitly don't cover the
GOTOInclusionTransformation.transformIndex(...) as it also was not
covered by any tests before and is never used. Furthermore, the entire
logic used to actually calculate new positions is already covered by the
added tests.

Adds test cases even if there already are existing tests covering the
same case. I decided to replace these tests instead of skipping such
cases for the new tests to be able to create an ordered, unified range
of tests covering all cases. The additional, now unnecessary test will
be removed in a separate commit.

Removes the comment in GOTOInclusionInformationTest asking for more test
cases.

</details>

<details><summary><b>[JUNIT][CORE] Remove unnecessary old transformation tests</b></summary>
<br>

Removes old inclusion transformation tests that are now contained in the
new range of tests.

I decided to replace these tests instead if just skipping the contained
cases in the new tests to be able to create an ordered, unified range of
tests covering all cases.

</details>

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 14
           

Complexity increasing per file
==============================
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/editor/text/TextPositionUtility.java  6
- core/test/junit/saros/test/util/OperationHelper.java  1
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/test/junit/saros/editor/text/TextPositionUtilityTest.java  5
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/test/junit/saros/test/util/OperationHelper.java  1
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/editor/text/TextPositionUtils.java  6
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- core/src/saros/activities/TextEditActivity.java  11
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/editor/text/TextPositionUtils.java  6
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Added test additional test cases for `SplitOperation`.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/editor/text/TextPositionUtils.java  6
- core/test/junit/saros/test/util/OperationHelper.java  1
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/editor/text/TextPositionUtils.java  6
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Added missing base case not covered by existing `GOTOInclusionTransformationTest`.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/src/saros/editor/text/TextPositionUtils.java  6
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Fix issue with `GOTOInclusionTransformation` handling found during test creation.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/activities/TextEditActivity.java  11
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/src/saros/editor/text/TextPositionUtils.java  6
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Rebased onto current base branch.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/editor/text/TextPositionUtils.java  6
- core/src/saros/activities/TextEditActivity.java  11
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/editor/text/TextPositionUtils.java  6
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/src/saros/activities/TextEditActivity.java  11
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Added additional multi-line tests for the inclusion transformation (`GOTOInclusionTransformation`).

Added an additional check to ensure that the applied operations during a `JupiterTest` actually have correct line and offset deltas.

Fixes several issues with the methods `GOTOInclusionTransformation.dropLeadingOverlap(DeleteOperation delA, DeleteOperation delB)` and `dropTrailingOverlap(DeleteOperation delA, DeleteOperation delB)` regarding their calculation of the new offset deltas and the new text in certain edge cases.

tobous commented:
So...yeah...this one got a liiiiiitle big. More than half of the changes are fixing or adding tests or test infrastructure, but it is still way to much to go through in a timely manner.

@m273d15 @srossbach How about talking the PR through together? That should make it easier to understand the changes. I am pretty open next week. Any interest?

m273d15 commented:
Sure, we can do it today

tobous commented on core/src/saros/activities/TextEditActivity.java:
```diff
@@ -22,67 +22,257 @@
+        + newTextOffsetDelta
         + ", new: '"
         + newText
+        + ", replaced text line delta: "
```
Add `'` and remove it from line 276.

tobous commented on core/test/junit/saros/concurrent/jupiter/test/util/Document.java:
```diff
@@ -85,7 +97,64 @@ public void execOperation(Operation op) {
+                + ", actual: "
+                + actualLineDelta
+                + "; operation: "
+                + text);
```
```suggestion
                + op);
```

tobous commented on core/test/junit/saros/concurrent/jupiter/test/util/Document.java:
```diff
@@ -85,7 +97,64 @@ public void execOperation(Operation op) {
+                + ", actual: "
+                + actualOffsetDelta
+                + "; operation: "
+                + text);
```
```suggestion
                + op);
```

m273d15 commented on core/test/junit/saros/concurrent/SplitOperationTest.java:
```diff
@@ -58,6 +57,93 @@ public void testInsertInsert() {
+  }
+
+  @Test
+  public void testMultiLineInsertInsertCase2() {
```
It is not really related to this test case, but what happens if a Saros user writes a "\n" in his/her shared document? Is this detected as linebreak or is it escaped?

tobous commented on core/src/saros/activities/TextEditActivity.java:
```diff
@@ -22,67 +22,257 @@
 
+  private final TextPosition startPosition;
+
+  private final int newTextLineDelta;
```
As far as I can tell, text edit activities are never actually marshaled/send. They are only used locally.
Instead, we transform them to Operations (DelteOperation, InsertOperation, SplitOperation) and send those. This seems sensible as we need the Operations for sending/receiving to apply the inclusion transformation/Jupiter algorithm.

tobous commented on core/src/saros/activities/TextEditActivity.java:
```diff
@@ -22,67 +22,257 @@
+   * @see TextPositionUtils#guessLineSeparator(String)
+   */
+  // TODO unify once content is normalized to Unix line separators
+  public TextEditActivity(
```
This is only meant as an ease of use CTOR.

Would introducing a static factory method or a factory class be a better solution in your opinion? (If so, which one?)

srossbach replied:
Static method is fine.

tobous commented on core/src/saros/editor/text/TextPositionUtils.java:
```diff
@@ -0,0 +1,160 @@
+   * @see #WINDOWS_LINE_SEPARATOR
+   * @see #UNIX_LINE_SEPARATOR
+   */
+  public static String guessLineSeparator(String text) {
```
From what I have heard, MacOS X hasn't been using `\r` as a line separator for a while now. They moved to being POSIX compliant a while ago and have been using UNIX line endings since then. 
(See corresponding [superuser entry](https://superuser.com/questions/439440/did-mac-os-lion-switch-to-using-line-feeds-lf-n-for-line-breaks-instead-of).)

tobous commented:
Adding `\n` to the file content gets escaped as `\\n`. These cases must be excluded from the logic looking for/counting new lines.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 14
           

Complexity increasing per file
==============================
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
- core/test/junit/saros/test/util/OperationHelper.java  1
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/editor/text/TextPositionUtils.java  8
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented on core/src/saros/activities/TextEditActivity.java:
```diff
@@ -110,55 +304,45 @@ public boolean equals(Object obj) {
+          new InsertOperation(startPosition, newTextLineDelta, newTextOffsetDelta, newText));
     }
 
-    // Cannot happen
```
Yeah, but I did not think the commented out code was very helpful. Can I just open an issue for this?

tobous replied:
Created issue #875 for this.

tobous commented on core/src/saros/concurrent/jupiter/internal/text/GOTOInclusionTransformation.java:
```diff
@@ -135,10 +135,13 @@ public int transformIndex(int index, Operation op, Object param) {
       }
     } else if (op instanceof DeleteOperation) {
       int pos = ((DeleteOperation) op).getPosition();
+      int length = ((DeleteOperation) op).getTextLength();
```
These logic is never used and was previously not covered by tests as well, as said in the [commit message](https://github.com/saros-project/saros/pull/872/commits/e6f4bfaff776172336bcdee6bd5dac85efa5f497).

I opened issue #870 for this. As long as it is not decided what to do with the methods, I don't think it is worth it to add tests. Furthermore, all of the actual calculations are already covered by other tests. The tests would only cover the if-else tree, i.e. that the right path is chosen for the right parameters.

tobous commented on core/test/junit/saros/concurrent/jupiter/test/puzzles/SimpleClientServerTest.java:
```diff
@@ -33,21 +33,21 @@ public void setupClientServer(String initialText) {
     setupClientServer("abc");
 
-    client.sendOperation(new InsertOperation(0, "e"), 100);
+    client.sendOperation(I(0, "e"), 100);
```
Because I only created the test helper class while migrating the tests, so I didn't want to go back and introduce it with the old logic and re-base the other changes on it.

But I can see how much work it takes to move it to a separate commit before the new position handling if you would prefer.

tobous replied:
Done.

tobous commented on core/src/saros/editor/text/TextPositionUtils.java:
```diff
@@ -0,0 +1,160 @@
+  public static final String UNIX_LINE_SEPARATOR = "\n";
+
+  /**
+   * Calculates the offset of the given text position in the given text using the given line
```
Wrong C&P. Will fix.

tobous replied:
Done.

tobous commented on core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java:
```diff
@@ -44,87 +52,170 @@
   @XStreamConverter(UrlEncodingStringConverter.class)
```
No, I did not even look at it. It was pre-existing logic and did not seem important for the purpose of this migration/refactoring to me.

tobous commented on core/src/saros/concurrent/jupiter/internal/text/GOTOInclusionTransformation.java:
```diff
@@ -118,166 +121,630 @@ public Operation transform(Operation op1, Operation op2, Object param) {
 
   @Override
-  public int transformIndex(int index, Operation op, Object param) {
+  public TextPosition transformIndex(TextPosition textPosition, Operation op, Object param) {
```
I don't know how I could have done it in any other way for this class. Even if I had managed to introduce the new `TextEditActivity` and operation classes in a separate commit, the logic here still would have to have been adjusted/overhauled all at once to match the new positioning scheme.

tobous commented on core/src/saros/editor/text/TextPositionUtils.java:
```diff
@@ -0,0 +1,169 @@
+    if (lineNumber == 0) {
+      return position.getInLineOffset();
+
+    } else if (lineSeparator.equals("")) {
```
```suggestion
    } else if (lineSeparator.isEmpty()) {
```

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 13
           

Complexity increasing per file
==============================
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/editor/text/TextPositionUtils.java  8
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
- core/test/junit/saros/test/util/OperationHelper.java  1
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Moved addition of `OperationHelper` and the introduction of its usage to separate commits. Also addressed other review comments.

tobous commented:
Rebased onto current base branch without any interaction.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 13
           

Complexity increasing per file
==============================
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/activities/TextEditActivity.java  11
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
- core/src/saros/editor/text/TextPositionUtils.java  8
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/test/junit/saros/test/util/OperationHelper.java  1
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
I also tested the changes to the inclusion transformation extensively using the fuzzer added in #874 to ensure that there aren't any subtle issue caused by split operations or longer operation transformation chains.

tobous commented:
I feel like I have addressed all existing review request.

@srossbach @m273d15 Do you have any other issues with this PR?

m273d15 commented:
> I feel like I have addressed all existing review request.

You addressed all my requests :+1:

tobous commented:
@srossbach What about you? Still got any concerns/issues with this PR?

tobous commented:
Rebased onto current base branch, resolving some minor merge issues caused by the logger refactoring and adjustment of the Eclipse contribution annotation logic.

tobous commented on core/test/junit/saros/concurrent/jupiter/test/puzzles/GOTOInclusionTransformationTest.java:
```diff
@@ -248,35 +221,67 @@ public void testTotallyConsumedDeleteOperation() {
+   * CharSequence, int)}, this assumption is explicitly tested.
+   */
+  @Test
+  public void escapedLineBreaksSanityCheck() {
```
Remove and add comment to other tests checking the same case.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 12
           

Complexity increasing per file
==============================
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/src/saros/activities/TextEditActivity.java  11
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/editor/text/TextPositionUtils.java  8
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Removed duplicate entry in `TextEditActivity` hash calculation. Removed line ending escaping sanity check from `GOTOInclusionTransformationTest` and added a comment to the exiting tests checking the same thing instead.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 13
           

Complexity increasing per file
==============================
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/editor/text/TextPositionUtils.java  8
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Rebased onto current base branch without any interaction.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 13
           

Complexity increasing per file
==============================
- core/src/saros/activities/TextEditActivity.java  11
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
- core/src/saros/editor/text/TextPositionUtils.java  8
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/test/junit/saros/test/util/OperationHelper.java  1
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Rebased onto current base branch without any interaction.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 13
           

Complexity increasing per file
==============================
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/editor/text/TextPositionUtils.java  8
- core/test/junit/saros/test/util/OperationHelper.java  1
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)

tobous commented:
Rebased onto current master, adding a private NOP CTOR to `TextPositionUtils`.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 16
- Added 13
           

Complexity increasing per file
==============================
- core/src/saros/concurrent/jupiter/internal/text/InsertOperation.java  4
- eclipse/src/saros/feedback/SelectionCollector.java  2
- core/src/saros/activities/TextEditActivity.java  11
- core/src/saros/concurrent/jupiter/internal/text/SplitOperation.java  2
- core/test/junit/saros/concurrent/jupiter/test/util/Document.java  3
- core/test/junit/saros/test/util/OperationHelper.java  1
- core/test/junit/saros/editor/text/TextPositionUtilsTest.java  5
- core/src/saros/concurrent/jupiter/internal/text/DeleteOperation.java  2
- core/src/saros/editor/text/TextPositionUtils.java  8
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/editor/EditorAPI.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5169723&bid=16971005)
